pkg/reconciler/image/reconcile_build.go

```diff
@@ -76,53 +77,36 @@ func (c *Reconciler) reconcileBuild(ctx context.Context, image *buildapi.Image,
 }
 
 func noScheduledBuild(buildNeeded corev1.ConditionStatus, builder buildapi.BuilderResource, build *buildapi.Build, sourceResolver *buildapi.SourceResolver) corev1alpha1.Conditions {
-	if !builder.Ready() {
-		return corev1alpha1.Conditions{
-			{
-				Type:               corev1alpha1.ConditionReady,
-				Status:             corev1.ConditionFalse,
-				Reason:             buildapi.BuilderNotReady,
-				Message:            builderError(builder),
-				LastTransitionTime: corev1alpha1.VolatileTime{Inner: metav1.Now()},
-			},
-			builderCondition(builder),
-		}
-	}
-	if buildNeeded == corev1.ConditionUnknown {
-		message := "Build status unknown"
-		if !sourceResolver.Ready() {
-			message = fmt.Sprintf("SourceResolver %s is not ready", sourceResolver.GetName())
-		}
-		return corev1alpha1.Conditions{
-			{
-				Type:               corev1alpha1.ConditionReady,
-				Status:             corev1.ConditionUnknown,
-				Reason:             ResolverNotReadyReason,
-				Message:            message,
-				LastTransitionTime: corev1alpha1.VolatileTime{Inner: metav1.Now()},
-			},
-			builderCondition(builder),
-		}
+	ready := corev1alpha1.Condition{
+		Type:               corev1alpha1.ConditionReady,
+		Status:             corev1.ConditionUnknown,
+		LastTransitionTime: corev1alpha1.VolatileTime{Inner: metav1.Now()},
 	}
 
-	buildReason := UpToDateReason
-	buildMessage := "Last build succeeded"
-
-	if !build.Status.GetCondition(corev1alpha1.ConditionSucceeded).IsTrue() {
-		buildReason = BuildFailedReason
-		buildMessage = "Last build did not succeed"
+	switch {
+	case !builder.Ready():
+		ready.Status = corev1.ConditionFalse
+		ready.Reason = buildapi.BuilderNotReady
+		ready.Message = builderError(builder)
+	case buildNeeded == corev1.ConditionUnknown && !sourceResolver.Ready():
+		ready.Status = corev1.ConditionUnknown
+		ready.Reason = ResolverNotReadyReason
+		ready.Message = fmt.Sprintf("SourceResolver %s is not ready", sourceResolver.GetName())
+	case buildNeeded == corev1.ConditionUnknown && sourceResolver.Ready():
+		ready.Status = corev1.ConditionUnknown
+		ready.Reason = UnknownStateReason
+		ready.Message = "Build status unknown"
```

Comment:
Excitingly, this code isn't covered, as I changed the reason and no tests failed.

(I've been carrying this state around, but I'm not sure what it's supposed to represent...)

Reply:
@matthewmcnew is this related to that libgit2 bug where the source resolver hangs forever? I can't think of any scenarios that would result in status unknown when both builder and source resolver are ready.

Reply:
This was https://github.com/pivotal/kpack/blob/7857820a20481e34b338c345ee4640ada53bb195/pkg/reconciler/image/reconcile_build.go#L74-L88, which would have had a message of `""` and status `Unknown` with no `Reason`.

This is now explicit enough that at least it's obvious that we don't know how we got into this state.  😁

Reply:
Looking at the source, it appears that _today_ the only way to get to this state is when SourceResolver or Builder are not ready.  Of course, a future refactor could change that, so I'd like to have some sort of default "not okay" status.